api/src/main/java/org/asynchttpclient/request/body/generator/ReactiveStreamBodyGenerator.java

```diff
@@ -0,0 +1,154 @@
+            return (isCompleted) ? State.Stop : State.Continue;
+        }
+    }
+
+    private static class QueueBackedSubscriber implements Subscriber<ByteBuffer> {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(QueueBackedSubscriber.class);
+
+        private final BlockingQueue<ByteBuffer> queue;
+        private volatile Subscription subscription;
+        private volatile ByteBuffer lastElement;
+        
+        private final AtomicBoolean completed = new AtomicBoolean(false);
+
+        public QueueBackedSubscriber() {
+            // capacity is 2 because the queue may contain the last element, together with an EMPTY element. 
+            // The latter is used to signal that the stream has completed.
+            this.queue = new ArrayBlockingQueue<ByteBuffer>(/* capacity */ 2);
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (s == null) throw null;
+
+            // If someone has made a mistake and added this Subscriber multiple times, let's handle it gracefully
+            if (this.subscription != null) { 
+                s.cancel(); // Cancel the additional subscription
+            }
+            else {
+              this.subscription = s;
+              this.subscription.request(1);
+            }
+        }
+
+        @Override
+        public void onNext(ByteBuffer t) {
+            if (t == null) throw null;
+            queue.add(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (t == null) throw null;
+            LOGGER.debug("Error occurred while consuming body stream.", t);
+        }
+
+        @Override
+        public void onComplete() {
+            completed.set(true);
+            queue.add(EMPTY); // this is to avoid that `queue.take` in method `readChunk` could block indefinitely
+        }
+
+        boolean readChunk(ByteBuffer buffer) {
+            try {
+                // take another element only if previous one was fully consumed and stream was not completed 
+                if(lastElement == null) {
+                    lastElement = queue.take();
+
+                    // we request another element only if previous one was fully consumed
+                    if(completed.get()) subscription.cancel();
+                    else subscription.request(1);
+                }
+
+                // If we are at the end of the stream, then this condition must be true.
+                // In fact, if we have received an empty chunk, but the stream is not completed, then we handle it 
+                // as any other chunk.
+                if(lastElement == EMPTY && completed.get()) return true;
+
+                // don't overflow the passed buffer
+                if(buffer.remaining() <= lastElement.remaining()) {
+                    int bufferSize = buffer.remaining();
+                    buffer.put(lastElement.array(), lastElement.position(), bufferSize);
```

Comment:
Oh, is that possible? Any idea how I should handle that? (should I create the array myself, or fail?)